examples/nn.dx

```diff
@@ -0,0 +1,241 @@
+  zero = (zero, zero)
+
+instance [VSpace a, VSpace b] VSpace (a & b)
+  scaleVec = \ s (a, b) . (scaleVec s a, scaleVec s b)
+
+def Layer (inp:Type) (out:Type) (param:Type) : Type=
+  {forward: param -> inp -> out &
+   init: Key -> param
+  }
+
+def forward (l:Layer i o p) (p : p) (x : i): o =
+  (getAt #forward l) p x
+
+def init (l:Layer i o p) (k:Key)  : p  =
+  (getAt #init l) k
+
+
+' ## Layers 
+
+' Dense layer
+
+def Dense (a:Type) (b:Type) : Type =
+  Layer (a=>Float) (b=>Float) ((a=>b=>Float) & (b=>Float))
+
+def dense (a:Type) (b:Type) : Dense a b =
+  {
+   forward = (\(weight, bias) x .
+               for j. (bias.j + sum for i. weight.i.j * x.i)),
+   init = arb
+  }
+
+' CNN layer
+
+def CNN (h:Int) (w:Int) (inc:Type) (outc:Type) (kw:Int) (kh:Int) : Type =
+  (Layer (inc=>Fin h=>Fin w=>Float)
+         (outc=>Fin h=>Fin w=>Float)
+         ((outc=>inc=>Fin kh=>Fin kw=>Float) &
+          (outc=>Float)))
+
+def conv2d (x:inc=>(Fin h)=>(Fin w)=>Float)
+           (kernel:outc=>inc=>(Fin kh)=>(Fin kw)=>Float) :
+     outc=>(Fin h)=>(Fin w)=>Float =
+     for o i j.
+         (i', j') = (ordinal i, ordinal j)
+         case (i' + kh) < h && (j' + kw) < w of
+          True ->
+              sum for (ki, kj, inp).
+                  (di, dj) = (fromOrdinal (Fin h) (i' + (ordinal ki)),
+                              fromOrdinal (Fin w) (j' + (ordinal kj)))
+                  x.inp.di.dj * kernel.o.inp.ki.kj
+          False -> zero
+
+def cnn (h:Int) ?-> (w:Int) ?-> (inc:Type) (outc:Type) (kw:Int) (kh:Int) :
+    CNN h w inc outc kw kh =
+  {
+    forward = \(weight, bias) x. for o i j . (conv2d x weight).o.i.j + bias.o,
+    init = arb
+  }
+
+' Pooling 
+
+def split (x: m=>v) : n=>o=>v =
+    for i j. x.((ordinal (i,j))@m)
+            
+def imtile (x: a=>b=>v) : n=>o=>p=>q=>v =
+    for kw kh w h. (split (split x).w.kw).h.kh
+
+def meanpool (kh: Type) (kw: Type) (x : m=>n=> Float) : ( h=>w=> Float) =
+    out : (kh => kw => h => w => Float) = imtile x
+    mean for (i,j). out.i.j
+
+' ## Simple point classifier
+
+[k1, k2] = splitKey $ newKey 1
+x1 : Fin 100 => Float = arb k1
+x2 : Fin 100 => Float = arb k2
+y = for i. case ((x1.i > 0.0) && (x2.i > 0.0)) || ((x1.i < 0.0) && (x2.i < 0.0)) of
+  True -> 1
+  False -> 0
+xs = for i. [x1.i, x2.i]
+
+
+:html showPlot $ xycPlot x1 x2 $ for i. IToF y.i
+
+simple = \h1.
+  ndense1 = dense (Fin 2) h1
+  ndense2 = dense h1 (Fin 2)
+  {
+    forward = (\ (dense1, dense2) x.
+         x1' = forward ndense1 dense1 x
+         x1 = for i. relu x1'.i
+         logsoftmax $ forward ndense2 dense2 x1),
+    init = (\key.
+         [k1, k2] = splitKey key
+         (init ndense1 k1, init ndense2 k2))
+  }
+
+
+def train_class [VSpace p] (model: Layer a (b=>Float) p)
+                           (x: batch=>a)
+                           (y: batch=>b)
+                           (epochs : Type)
+                           (minibatch : Type)  (minibatches : Type) :
```

Comment:
What is the relation between the `Type` arguments `minibatch` and `minibatches`?

I guess I can find out from call sites - some doc comments explaining them would be great.